File: chartview/src/main/java/com/bitvale/chartview/ChartSpinner.java
package com.bitvale.chartview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.*;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import androidx.annotation.ColorInt;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Created by Alexander Kolpakov (jquickapp@gmail.com) on 16-Mar-19
 */
public class ChartSpinner extends View {

    private ChartSpinnerListener listener;

    private ArrayList<Long> xAxis;
    private ArrayList<Chart.Column> yAxis = new ArrayList<Chart.Column>();

    private Paint chartPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);

    @ColorInt
    private int foregroundColor = 0;
    @ColorInt
    private int frameColor = 0;

    private int spinnerHeight = 0;

    private Path chartPath = new Path();

    private float xMultiplier = 0f;
    private float yMultiplier = 0f;

    private float smallPadding = 0f;
    private int minFrameWidth = 0;
    private int frameSideSize = 0;
    private Rect frameOuterRect = new Rect();
    private Rect frameInnerRect = new Rect();
    private Region frameRegion = new Region();

    private int currentFrameWidth = 0;

    private float dX = 0f;
    private int currentOuterLeft = 0;
    private int currentOuterRight = 0;

    private boolean moveLeftBorder = false;
    private boolean moveRightBorder = false;

    public ChartSpinner(Context context) {
        super(context);
        init(context, null, 0);
    }

    public ChartSpinner(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs, 0);
    }

    public ChartSpinner(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        setId(generateViewId());
        setLayerType(LAYER_TYPE_SOFTWARE, null);

        TypedArray a = context.obtainStyledAttributes(
                attrs,
                R.styleable.ChartSpinner,
                R.attr.chartSpinnerStyle,
                R.style.ChartSpinner
        );

        foregroundColor = a.getColor(R.styleable.ChartSpinner_foreground_color, 0);
        frameColor = a.getColor(R.styleable.ChartSpinner_frame_color, 0);

        spinnerHeight = a.getDimensionPixelOffset(R.styleable.ChartSpinner_spinner_height, 0);

        chartPaint.setStyle(Paint.Style.STROKE);

        smallPadding = context.getResources().getDisplayMetrics().density * 4;
        a.recycle();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int widthSize = View.MeasureSpec.getSize(widthMeasureSpec);
        setMeasuredDimension(widthSize, spinnerHeight);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        if (yAxis.isEmpty()) return;
        calculateMultipliers();
        calculateFrameSize();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (yAxis.isEmpty()) return;

        for (int i = 0; i < yAxis.size(); i++) {
            if (yAxis.get(i).enabled) drawChart(canvas, yAxis.get(i));
        }
        drawFrame(canvas);
        drawForeground(canvas);
    }

    private void drawChart(Canvas canvas, Chart.Column column) {
        chartPaint.setColor(Color.parseColor(column.color));
        chartPaint.setStrokeWidth(4f);
        chartPath.reset();

        for (int i = 0; i < column.values.size(); i++) {
            float x = i * xMultiplier;
            float y = getHeight() - column.values.get(i) * yMultiplier;
            if (y >= getHeight() / 2) y -= smallPadding;
            else y += smallPadding;
            if (i == 0) {
                chartPath.moveTo(x, y);
            } else {
                chartPath.lineTo(x, y);
            }
        }
        canvas.drawPath(chartPath, chartPaint);
    }

    private void drawFrame(Canvas canvas) {
        paint.setColor(frameColor);
        frameRegion.set(frameOuterRect);
        frameRegion.op(frameInnerRect, Region.Op.XOR);
        canvas.drawPath(frameRegion.getBoundaryPath(), paint);
    }

    private void calculateFrameSize() {
        minFrameWidth = 6 * (getWidth() / xAxis.size());
        frameSideSize = minFrameWidth / 4;
        if (currentFrameWidth == 0) currentFrameWidth = minFrameWidth;
        updateFrameSize();
    }

    private void updateFrameSize() {
        frameOuterRect.set(getWidth() - currentFrameWidth, 0, getWidth(), getHeight());
        frameInnerRect.set(
                (int) (frameOuterRect.left + (frameSideSize * 1.5)),
                (frameOuterRect.top + frameSideSize / 2),
                (int) (frameOuterRect.right - (frameSideSize * 1.5)),
                (frameOuterRect.bottom - frameSideSize / 2)
        );
        calculateChartData(moveRightBorder, moveLeftBorder);
    }

    private void calculateChartData(boolean moveRightBorder, boolean moveLeftBorder) {
        int frameStartPosition = frameOuterRect.left;
        int frameEndPosition = frameOuterRect.right;
        int frameWidth = frameEndPosition - frameStartPosition;
        float oneDayWidth = getWidth() / ((float) xAxis.size());
        int daysInFrame = (int) (Math.ceil(frameWidth / oneDayWidth));
        int daysAfterFrame = (int) (Math.floor((getWidth() - frameEndPosition) / oneDayWidth));
        int daysBeforeFrame = (int) (Math.floor((frameStartPosition / oneDayWidth)));

        if (moveLeftBorder) {
            if (daysBeforeFrame < xAxis.size() - daysAfterFrame - daysInFrame) daysBeforeFrame =
                    xAxis.size() - daysAfterFrame - daysInFrame;
        }
        if (moveRightBorder) {
            if (daysAfterFrame > xAxis.size() - daysBeforeFrame - daysInFrame) daysAfterFrame =
                    xAxis.size() - daysAfterFrame - daysInFrame;
        }
        listener.onRangeChanged(daysBeforeFrame, daysAfterFrame, daysInFrame);
    }

    private void drawForeground(Canvas canvas) {
        paint.setColor(foregroundColor);
        canvas.drawRect(0f, 0f, frameOuterRect.left, spinnerHeight, paint);
        canvas.drawRect(frameOuterRect.right, 0f, getWidth(), spinnerHeight, paint);
    }

    void setupData(Chart chart) {
        xAxis = chart.columns.get(0).values;
        for (int i = 1; i < chart.columns.size(); i++) {
            yAxis.add(chart.columns.get(i));
        }
        invalidate();
    }

    private void calculateMultipliers() {
        int max = 0;
        for (int i = 0; i < yAxis.size(); i++) {
            long newMax = Collections.max(yAxis.get(i).values);
            if (newMax > max) max = (int) newMax;
        }

        xMultiplier = ((float) getWidth()) / (xAxis.size() - 1);
        yMultiplier = ((float) spinnerHeight) / max;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x = event.getX();
        int action = event.getAction() & MotionEvent.ACTION_MASK;

        switch (action) {
            case MotionEvent.ACTION_DOWN:
                dX = x;
                currentOuterLeft = frameOuterRect.left;
                currentOuterRight = frameOuterRect.right;

                moveLeftBorder = false;
                moveRightBorder = false;

                if (x < frameInnerRect.left + 20) {
                    moveLeftBorder = true;
                }
                if (x > frameInnerRect.right - 20) {
                    moveRightBorder = true;
                }
                break;
            case MotionEvent.ACTION_MOVE:
                int l = currentOuterLeft + (int) (x - dX);
                int r = currentOuterRight + (int) (x - dX);
                if (moveLeftBorder) {
                    frameOuterRect.left = l;
                    if (frameOuterRect.width() < minFrameWidth) {
                        frameOuterRect.left = frameOuterRect.right - minFrameWidth;
                    }
                    if (frameOuterRect.left < 0) {
                        frameOuterRect.left = 0;
                    }
                    frameInnerRect.left = frameOuterRect.left + (int) (frameSideSize * 1.5);
                } else {
                    if (moveRightBorder) {
                        frameOuterRect.right = r;
                        if (frameOuterRect.width() < minFrameWidth) {
                            frameOuterRect.right = frameOuterRect.left + minFrameWidth;
                        }
                        if (frameOuterRect.right > getWidth()) {
                            frameOuterRect.right = getWidth();
                        }
                        frameInnerRect.right = frameOuterRect.right - (int) (frameSideSize * 1.5);
                    } else {
                        if (l < 0) {
                            r = frameOuterRect.right;
                            l = 0;
                        }
                        if (r > getWidth()) {
                            r = getWidth();
                            l = frameOuterRect.left;
                        }
                        frameOuterRect.left = l;
                        frameOuterRect.right = r;
                        frameInnerRect.left = l + ((int) (frameSideSize * 1.5));
                        frameInnerRect.right = r - ((int) (frameSideSize * 1.5));
                    }
                }
                break;
        }
        calculateChartData(moveRightBorder, moveLeftBorder);
        invalidate();
        return true;
    }


    public void setChartListener(ChartSpinnerListener listener) {
        this.listener = listener;
    }
}

File: chartview/src/main/java/com/bitvale/chartview/widget/chart/ChartSpinner.kt
package com.bitvale.chartview.widget.chart

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.annotation.ColorInt
import androidx.core.content.res.getColorOrThrow
import androidx.core.content.res.getDimensionPixelOffsetOrThrow
import com.bitvale.chartview.model.Chart
import com.bitvale.chartview.ChartSpinnerListener
import com.bitvale.chartview.R


/**
 * Created by Alexander Kolpakov (jquickapp@gmail.com) on 10-Mar-19
 */
class ChartSpinner @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private var listener: ChartSpinnerListener? = null

    private var xAxis = ArrayList<Long>()
    private val yAxis = ArrayList<Chart.Column>()

    private val chartPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)

    @ColorInt
    private var foregroundColor: Int = 0
    @ColorInt
    private var frameColor: Int = 0

    private var spinnerHeight: Int = 0

    private val chartPath = Path()

    private var xMultiplier = 0f
    private var yMultiplier = 0f

    private var smallPadding = 0f
    private var minFrameWidth = 0
    private var frameSideSize = 0
    private val frameOuterRect = Rect()
    private val frameInnerRect = Rect()
    private val frameRegion = Region()

    private var currentFrameWidth = 0

    private var dX: Float = 0f
    private var deltaX: Float = 0f

    var currentOuterLeft = 0
    var currentOuterRight = 0

    var moveLeftBorder = false
    var moveRightBorder = false
    init {
        id = generateViewId()
        setLayerType(LAYER_TYPE_SOFTWARE, null)
        attrs?.let { retrieveAttributes(attrs) }
    }

    private fun retrieveAttributes(attrs: AttributeSet) {
        val a = context.obtainStyledAttributes(
            attrs,
            R.styleable.ChartSpinner,
            R.attr.chartSpinnerStyle,
            R.style.ChartSpinner
        )

        foregroundColor = a.getColorOrThrow(R.styleable.ChartSpinner_foreground_color)
        frameColor = a.getColorOrThrow(R.styleable.ChartSpinner_frame_color)

        spinnerHeight = a.getDimensionPixelOffsetOrThrow(R.styleable.ChartSpinner_spinner_height)

        chartPaint.apply {
            style = Paint.Style.STROKE
        }

        smallPadding = resources.displayMetrics.density * 4
        a.recycle()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val widthSize = View.MeasureSpec.getSize(widthMeasureSpec)
        setMeasuredDimension(widthSize, spinnerHeight)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (yAxis.isEmpty()) return
        calculateMultipliers()
        calculateFrameSize()
    }

    override fun onDraw(canvas: Canvas?) {
        if (yAxis.isEmpty()) return
        yAxis.forEach {
            if (it.enabled) drawChart(canvas, it)
        }
        drawFrame(canvas)
        drawForeground(canvas)
    }

    private fun drawChart(canvas: Canvas?, column: Chart.Column) {
        chartPaint.color = Color.parseColor(column.color)
        chartPaint.strokeWidth = 4f
        chartPath.reset()
        var i = 0
        column.values.forEach {
            val x = i * xMultiplier
            var y = height - it * yMultiplier
            if (y >= height / 2) y -= smallPadding
            else y += smallPadding
            if (i == 0) {
                chartPath.moveTo(x, y)
            } else {
                chartPath.lineTo(x, y)
            }
            i++
        }

        canvas?.drawPath(chartPath, chartPaint)
    }

    private fun drawFrame(canvas: Canvas?) {
        paint.color = frameColor
        frameRegion.set(frameOuterRect)
        frameRegion.op(frameInnerRect, Region.Op.XOR)
        canvas?.drawPath(frameRegion.boundaryPath, paint)
    }

    private fun calculateFrameSize() {
        minFrameWidth = 6 * (width / xAxis.size)
        frameSideSize = minFrameWidth / 4
        if (currentFrameWidth == 0) currentFrameWidth = minFrameWidth
        updateFrameSize()
    }

    private fun updateFrameSize() {
        frameOuterRect.set(width - currentFrameWidth, 0, width, height)
        frameInnerRect.set(
            frameOuterRect.left + (frameSideSize * 1.5).toInt(),
            frameOuterRect.top + frameSideSize / 2,
            frameOuterRect.right - (frameSideSize * 1.5).toInt(),
            frameOuterRect.bottom - frameSideSize / 2
        )
        calculateChartData(moveRightBorder, moveLeftBorder)
    }

    private fun calculateChartData(moveRightBorder: Boolean, moveLeftBorder: Boolean) {
        val frameStartPosition = frameOuterRect.left
        val frameEndPosition = frameOuterRect.right
        val frameWidth = frameEndPosition - frameStartPosition
        val oneDayWidth = width / xAxis.size.toFloat()
        val daysInFrame = Math.ceil((frameWidth / oneDayWidth).toDouble()).toInt()
        var daysAfterFrame = Math.floor(((width - frameEndPosition) / oneDayWidth).toDouble()).toInt()
        var daysBeforeFrame = Math.floor((frameStartPosition / oneDayWidth).toDouble()).toInt()

        if (moveLeftBorder) {
            if (daysBeforeFrame < xAxis.size - daysAfterFrame - daysInFrame) daysBeforeFrame =
                xAxis.size - daysAfterFrame - daysInFrame
        }
        if (moveRightBorder) {
            if (daysAfterFrame > xAxis.size - daysBeforeFrame - daysInFrame) daysAfterFrame =
                xAxis.size - daysAfterFrame - daysInFrame
        }
        listener?.onRangeChanged(daysBeforeFrame, daysAfterFrame, daysInFrame, deltaX)
    }

    private fun drawForeground(canvas: Canvas?) {
        paint.color = foregroundColor
        canvas?.drawRect(0f, 0f, frameOuterRect.left.toFloat(), spinnerHeight.toFloat(), paint)
        canvas?.drawRect(frameOuterRect.right.toFloat(), 0f, width.toFloat(), spinnerHeight.toFloat(), paint)
    }

    private fun calculateMultipliers() {
        var max = 0
        yAxis.forEach {
            val newMax = it.values.max() ?: 0
            if (newMax > max) max = newMax.toInt()
        }

        xMultiplier = width.toFloat() / (xAxis.size - 1)
        yMultiplier = spinnerHeight.toFloat() / max
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x
        val action = event.action and MotionEvent.ACTION_MASK

        when (action) {
            MotionEvent.ACTION_DOWN -> {
                dX = x
                currentOuterLeft = frameOuterRect.left
                currentOuterRight = frameOuterRect.right

                moveLeftBorder = false
                moveRightBorder = false

                if (x < frameInnerRect.left + 20) {
                    moveLeftBorder = true
                }
                if (x > frameInnerRect.right - 20) {
                    moveRightBorder = true
                }
            }
            MotionEvent.ACTION_MOVE -> {
                deltaX = 0f
                var l = currentOuterLeft + (x - dX).toInt()
                var r = currentOuterRight + (x - dX).toInt()
                if (moveLeftBorder) {
                    frameOuterRect.left = l
                    if (frameOuterRect.width() < minFrameWidth) {
                        frameOuterRect.left = frameOuterRect.right - minFrameWidth
                    }
                    if (frameOuterRect.left < 0) {
                        frameOuterRect.left = 0
                    }
                    frameInnerRect.left = frameOuterRect.left + (frameSideSize * 1.5).toInt()
                } else {
                    if (moveRightBorder) {
                        frameOuterRect.right = r
                        if (frameOuterRect.width() < minFrameWidth) {
                            frameOuterRect.right = frameOuterRect.left + minFrameWidth
                        }
                        if (frameOuterRect.right > width) {
                            frameOuterRect.right = width
                        }
                        frameInnerRect.right = frameOuterRect.right - (frameSideSize * 1.5).toInt()
                    } else {
                        deltaX = x - dX
                        if (l < 0) {
                            r = frameOuterRect.right
                            l = 0
                        }
                        if (r > width) {
                            r = width
                            l = frameOuterRect.left
                        }
                        frameOuterRect.left = l
                        frameOuterRect.right = r
                        frameInnerRect.left = l + (frameSideSize * 1.5).toInt()
                        frameInnerRect.right = r - (frameSideSize * 1.5).toInt()
                    }
                }
            }
        }
        calculateChartData(moveRightBorder, moveLeftBorder)
        invalidate()
        return true
    }

    fun setChartListener(listener: ChartSpinnerListener) {
        this.listener = listener
    }

    fun setupData(xAxis: ArrayList<Long>, yAxis: ArrayList<Chart.Column>) {
        this.xAxis.clear()
        this.xAxis.addAll(xAxis)
        this.yAxis.clear()
        this.yAxis.addAll(yAxis)
        invalidate()
    }
}